File: src/main/java/uk/yermak/audiobookconverter/ConversionContext.java
package uk.yermak.audiobookconverter;

import com.google.gson.Gson;
import javafx.application.Platform;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.invoke.MethodHandles;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

/**
 * Created by yermak on 06-Feb-18.
 */
public class ConversionContext {

    final static Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

    private final LinkedList<ConversionJob> conversionQueue = new LinkedList<>();
    private final SimpleObjectProperty<ConversionGroup> conversionGroupHolder = new SimpleObjectProperty<>(new ConversionGroup());
    private boolean paused;

    private final ObservableList<MediaInfo> selectedMedia = FXCollections.observableArrayList();
    private final GenresManager genresManager = new GenresManager();

/*
    private final SimpleObjectProperty<AudioBookInfo> bookInfo = new SimpleObjectProperty<>();
    private final SimpleObjectProperty<Book> book = new SimpleObjectProperty<>();
    private final ObservableList<MediaInfo> media = FXCollections.observableArrayList();
    private final ObservableList<ArtWork> posters = FXCollections.observableArrayList();
    private final SimpleObjectProperty<OutputParameters> outputParameters = new SimpleObjectProperty<>(Preset.DEFAULT_OUTPUT_PARAMETERS);
    private final SimpleObjectProperty<Double> speed = new SimpleObjectProperty<>(1.0);
*/

    private final static ExecutorService executorService = Executors.newCachedThreadPool();


    public ConversionContext() {
        resetForNewConversion();
    }


    public void saveGenres() {
        genresManager.saveGenres(conversionGroupHolder.get().getBookInfo());
    }

    public ObservableList<String> loadGenres() {
        return genresManager.loadGenres();
    }

    public void resetForNewConversion() {
        saveGenres();
        ConversionGroup newConversionGroup = new ConversionGroup();
        conversionGroupHolder.set(newConversionGroup);
//        newConversionGroup.setBookInfo(AudioBookInfo.instance());
/*
        bookInfo.set(AudioBookInfo.instance());
        book.set(null);
        posters.clear();
        media.clear();
*/
    }

    public void stopConversions() {
        conversionQueue.forEach(ConversionJob::stop);
    }

    public ObservableList<MediaInfo> getSelectedMedia() {
        return selectedMedia;
    }

    public void pauseConversions() {
        conversionQueue.forEach(ConversionJob::pause);
        paused = true;
    }

    public void resumeConversions() {
        conversionQueue.forEach(ConversionJob::resume);
        paused = false;
    }

    public boolean isPaused() {
        return paused;
    }

    public ConversionGroup next() {
        return conversionGroupHolder.get();
    }

    public void addJob(ConversionJob conversionJob) {
        conversionQueue.add(conversionJob);
        executorService.execute(conversionJob);
    }


}

File: src/main/scala/uk/yermak/audiobookconverter/ConversionContext.scala
package uk.yermak.audiobookconverter

import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.{ChangeListener, ObservableValue}
import javafx.collections.FXCollections
import uk.yermak.audiobookconverter.fx.ConversionProgress

import scala.collection.mutable.ListBuffer

/**
  * Created by yermak on 06-Feb-18.
  */
class ConversionContext() {

  private val conversionQueue: ListBuffer[Conversion] = ListBuffer()
  private val conversion: SimpleObjectProperty[Conversion] = new SimpleObjectProperty[Conversion](new Conversion)
  private var paused = false
  private var subscriber: Subscriber = _
  private val selectedMedia = FXCollections.observableArrayList[MediaInfo]
  conversionQueue += conversion.get

  //TODO simplify Subscriber
  def startConversion(outputDestination: String, conversionProgress: ConversionProgress) = {
    subscriber.addConversionProgress(conversionProgress)
    conversion.get.start(outputDestination, conversionProgress)
    val newConversion = new Conversion
    conversionQueue += newConversion
    conversion.set(newConversion)
  }

  def stopConversions() = conversionQueue.foreach(_.stop)

  def subscribeForStart(subscriber: Subscriber) = this.subscriber = subscriber

  def getSelectedMedia = selectedMedia

  def registerForConversion(conversionSubscriber: ConversionSubscriber): Conversion = {
    conversion.addListener(new ConversionChangeListener(conversionSubscriber))
    conversion.get
  }

  def pauseConversions() = {
    conversionQueue.foreach(_.pause)
    paused = true
  }

  def resumeConversions() = {
    conversionQueue.foreach(_.resume)
    paused = false
  }

  def isPaused = paused
}

class ConversionChangeListener(conversionSubscriber: ConversionSubscriber) extends ChangeListener[Conversion] {

  override def changed(observable: ObservableValue[_ <: Conversion], oldValue: Conversion, newValue: Conversion) = {
    conversionSubscriber.resetForNewConversion(newValue)
  }
}